kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/storageApi/StorageWriteApiBatchApplicationStream.java

```diff
@@ -0,0 +1,302 @@
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
+/**
+ * An extension of {@link StorageWriteApiApplicationStream} which uses application streams for batch loading data
+ * following at least once semantic
+ * Current/Active stream means - Stream which is not yet finalised and would be used for any new data append
+ * Other streams (non-current/ non-active streams) - These streams may/may not be finalised yet but would not be used
+ * for any new data. These will only write data for offsets assigned so far.
+ */
+public class StorageWriteApiBatchApplicationStream extends StorageWriteApiApplicationStream {
+
+    private static final Logger logger = LoggerFactory.getLogger(StorageWriteApiBatchApplicationStream.class);
+
+    /**
+     * Map of <tableName , <StreamName, {@link ApplicationStream}>>
+     * Streams should be accessed in the order of entry, so we need LinkedHashMap here
+     */
+    protected ConcurrentMap<String, LinkedHashMap<String, ApplicationStream>> streams;
+
+    /**
+     * Quick lookup for current open stream by tableName
+     */
+    protected ConcurrentMap<String, String> currentStreams;
+
+    /**
+     * Lock on table names to prevent execution of critical section by multiple threads
+     */
+    protected ConcurrentMap<String, Object> tableLocks;
+
+    public StorageWriteApiBatchApplicationStream(
+            int retry,
+            long retryWait,
+            BigQueryWriteSettings writeSettings,
+            boolean autoCreateTables,
+            ErrantRecordHandler errantRecordHandler,
+            SchemaManager schemaManager) {
+        super(retry, retryWait, writeSettings, autoCreateTables, errantRecordHandler, schemaManager);
+        streams = new ConcurrentHashMap<>();
+        currentStreams = new ConcurrentHashMap<>();
+        tableLocks = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Takes care of resource cleanup
+     */
+    @Override
+    public void preShutdown() {
+        logger.debug("Shutting down all streams on all tables as due to task shutdown!!!");
+        this.streams.values()
+                .stream().flatMap(item -> item.values().stream())
+                .collect(Collectors.toList())
+                .forEach(ApplicationStream::closeStream);
+        logger.debug("Shutting completed for all streams on all tables!");
+    }
+
+    /**
+     * Calls storage Api's append
+     *
+     * @param tableName  The table to write data to in project/dataset/tableName format
+     * @param rows       The records to write
+     * @param streamName The stream to use to write table to table.
+     */
+    @Override
+    public void appendRows(TableName tableName, List<Object[]> rows, String streamName) {
+        JSONArray jsonArray = new JSONArray();
```

Comment:
Minor - could we rename this variable to something it is actually storing in the array? Will improve code readability.